### DISCUSSION TITLE
useOnViewportChange callbacks don't always trigger

### DISCUSSION THREAD

**Discussion by jukeboxhero123** [xyflow/xyflow#3184] [General]
I'm using the useOnViewportChange hook similarly to the example on the documentation shown below:

```
import { useCallback } from 'react';
import ReactFlow, { useOnViewportChange } from 'reactflow';

function ViewportChangeLogger() {
  useOnViewportChange({
    onStart: useCallback((viewport: Viewport) => console.log('start', viewport), []),
    onChange: useCallback((viewport: Viewport) => console.log('change', viewport), []),
    onEnd: useCallback((viewport: Viewport) => console.log('end', viewport), []),
  });

  return null;
}

```
I'm finding that when viewport changes are done by zooming/panning with the mouse, all events will fire. But this doesn't seem to be the case in other viewport changing methods. For example, when changing the viewport via the Controls, the onStart and onEnd callbacks do not fire while the onChange does. I'm wondering why this might be the case? If it's intentional, is there anyway to get around this?

**Comment by bcakmakoglu** [comment]
I would think it's the expected behavior because why would a button press call `onStart` and `onEnd`?
The button press does a direct change, while using the mouse has a time related component to it (how long and what direction are you scrolling for to zoom/pan determines how far in/out you zoom) so logically it makes sense that this can have a *start* and and *end*.

At least that's my thoughts on this 🤷‍♂️
